![Ironhack logo](https://i.imgur.com/1QgrNNw.png)

<body>
    <p style="font-size:28px;text-align:center"><b>Project 02 | Data Cleaning & Manipulation</b></p>
</body>

# Introduction

The objective of this project was to answer a problem statement, practicing data cleaning and manipulation.

---

<body>
    <p style="font-size:20px"><b>Problem Statement</b></p>
</body>

_What are the most common characteristics of people involved in shark incidents in the history?_

---

To answer this problem, the following characteristics will be analyzed:
- Gender
- Age
- Activity

# Setup

## Import

In [1]:
import pandas as pd
import numpy as np

## Load the dataset

In [2]:
# Load the dataset from a Excel file into a Pandas DataFrame
df = pd.read_excel('GSAF5.xls')

In [3]:
# Check information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25792 entries, 0 to 25791
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8771 non-null   object 
 1   Date                    6558 non-null   object 
 2   Year                    6556 non-null   float64
 3   Type                    6552 non-null   object 
 4   Country                 6508 non-null   object 
 5   Area                    6091 non-null   object 
 6   Location                6008 non-null   object 
 7   Activity                6002 non-null   object 
 8   Name                    6343 non-null   object 
 9   Sex                     5987 non-null   object 
 10  Age                     3660 non-null   object 
 11  Injury                  6528 non-null   object 
 12  Fatal (Y/N)             6006 non-null   object 
 13  Time                    3139 non-null   object 
 14  Species                 3610 non-null 

### Conclusions about the dataset for cleaning purposes
1. Seeing the information about the dataset, it is clear that there are a lot of rows that only contains missing values `NaN`.
2. The columns 'Unnamed: 22' and 'Unnamed: 23' probably do not have any information since they have 1 and 2 non-null values, respectively.

## Create a backup version of the raw dataset

In [4]:
# Create a backup version
df_raw_bkp = df.copy()

# General Data Cleaning

## Headers

In [5]:
# Remove unnecessary spaces, put everything in lower case and replace spaces with underscores
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')
df.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species', 'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

In [6]:
# Rename the column 'fatal_(y/n)' to make it simpler
df = df.rename(columns={'fatal_(y/n)': 'fatal'})
df.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal', 'time', 'species',
       'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

## Columns that are not relevant to the analysis

As mentioned above, the columns `unnamed:_22` and `unnamed:_23` probably do not have any relevant information, so they can be removed.

In [7]:
# Remove the columns 'unnamed:_22' and 'unnamed:_23'
df = df.drop(columns=['unnamed:_22', 'unnamed:_23'])

# Check the result
df.head()

case_number         date    year        type    country             area  \
0    2020.08.20  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1    2020.08.14  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2    2020.08.10  10-Aug-2020  2020.0    Provoked        USA          Florida   
3    2020.08.02  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  2020.07.31.c  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity             name  \
0         New Smyrna Beach, Volusia County  Boogie boarding   Carolina Jones   
1             Shelly Beach, Port Macquarie          Surfing  Chantelle Doyle   
2  Off Gasparilla Island, Charlotte County          Fishing             male   
3                   Candle Reef, St. Croix       Snorkeling     Melony Klein   
4         New Smyrna Beach, Volusia County          Surfing      Megan Tossi   

  sex  ... fatal   time               species  \
0   F  ...     N  11h00                   NaN   
1   F  ...     N  09h30  White shark, 2-to 3m   
2   M  ...     N  16h00    Blacktip shark, 6'   
3   F  ...     N  14h00       Nurse shark, 5'   
4   F  ...     N  17h00                   NaN   

            investigator_or_source                         pdf  \
0  K. McMurray, TrackingSharks.com          2020.08.20-NSB.pdf   
1                   B. Myatt, GSAF  2020.08.14-ShellyBeach.pdf   
2  K. McMurray, TrackingSharks.com     2020.08.10-Provoked.pdf   
3  K. McMurray, TrackingSharks.com        2020.08.02-Klein.pdf   
4  K. McMurray, TrackingSharks.com     2020.07.31.c-Tossi..pdf   

                                        href_formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href case_number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2020.08.20   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2020.08.14   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2020.08.10   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2020.08.02   
4  http://sharkattackfile.net/spreadsheets/pdf_di...  2020.07.31.c   

  case_number.2 original_order  
0    2020.08.20         6559.0  
1    2020.08.14         6558.0  
2    2020.08.10         6557.0  
3    2020.08.02         6556.0  
4  2020.07.31.c         6555.0  

[5 rows x 22 columns]

Checking the table above, it is clear that the columns `investigator_or_source`, `pdf`, `href_formula` and `href` will not be relevant to the analysis since they are just information about the source of each incident. Therefore, these columns can be removed.

In [8]:
# Remove the columns 'nvestigator_or_source', 'pdf', 'href_formula' and 'href'
df = df.drop(columns=['investigator_or_source', 'pdf', 'href_formula', 'href'])

# Check the result
df.head()

case_number         date    year        type    country             area  \
0    2020.08.20  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1    2020.08.14  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2    2020.08.10  10-Aug-2020  2020.0    Provoked        USA          Florida   
3    2020.08.02  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  2020.07.31.c  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity             name  \
0         New Smyrna Beach, Volusia County  Boogie boarding   Carolina Jones   
1             Shelly Beach, Port Macquarie          Surfing  Chantelle Doyle   
2  Off Gasparilla Island, Charlotte County          Fishing             male   
3                   Candle Reef, St. Croix       Snorkeling     Melony Klein   
4         New Smyrna Beach, Volusia County          Surfing      Megan Tossi   

  sex  age                                             injury fatal   time  \
0   F   50                      Minor lacerations to left leg     N  11h00   
1   F   35      Lacerations to right calf and posterior thigh     N  09h30   
2   M   55  Injury to left forearm by hooked shark PROVOKE...     N  16h00   
3   F  NaN                      Lacerations to hand and wrist     N  14h00   
4   F   22                               Lacerations to foot      N  17h00   

                species case_number.1 case_number.2  original_order  
0                   NaN    2020.08.20    2020.08.20          6559.0  
1  White shark, 2-to 3m    2020.08.14    2020.08.14          6558.0  
2    Blacktip shark, 6'    2020.08.10    2020.08.10          6557.0  
3       Nurse shark, 5'    2020.08.02    2020.08.02          6556.0  
4                   NaN  2020.07.31.c  2020.07.31.c          6555.0

The content of columns `case_number`, `case_number.1`, `case_number.2` and `original_order` is just a way to reference each case and is not relevant to the analysis. So, they will also be removed.

In [9]:
# Remove the columns 'case_number', 'case_number.1', 'case_number.2' and 'original_order'
df = df.drop(columns=['case_number', 'case_number.1', 'case_number.2', 'original_order'])

# Check the result
df.head()

date    year        type    country             area  \
0  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2  10-Aug-2020  2020.0    Provoked        USA          Florida   
3  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity             name  \
0         New Smyrna Beach, Volusia County  Boogie boarding   Carolina Jones   
1             Shelly Beach, Port Macquarie          Surfing  Chantelle Doyle   
2  Off Gasparilla Island, Charlotte County          Fishing             male   
3                   Candle Reef, St. Croix       Snorkeling     Melony Klein   
4         New Smyrna Beach, Volusia County          Surfing      Megan Tossi   

  sex  age                                             injury fatal   time  \
0   F   50                      Minor lacerations to left leg     N  11h00   
1   F   35      Lacerations to right calf and posterior thigh     N  09h30   
2   M   55  Injury to left forearm by hooked shark PROVOKE...     N  16h00   
3   F  NaN                      Lacerations to hand and wrist     N  14h00   
4   F   22                               Lacerations to foot      N  17h00   

                species  
0                   NaN  
1  White shark, 2-to 3m  
2    Blacktip shark, 6'  
3       Nurse shark, 5'  
4                   NaN

After removing the unnecessary columns, the information about the new dataframe is shown below.

The column `name` is also irrelevant, since there will not be a specific analysis of each individual. So, this column can be removed.

In [10]:
# Remove column 'name'
df = df.drop(columns='name')

# Check the result
df.head()

date    year        type    country             area  \
0  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2  10-Aug-2020  2020.0    Provoked        USA          Florida   
3  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity sex  age  \
0         New Smyrna Beach, Volusia County  Boogie boarding   F   50   
1             Shelly Beach, Port Macquarie          Surfing   F   35   
2  Off Gasparilla Island, Charlotte County          Fishing   M   55   
3                   Candle Reef, St. Croix       Snorkeling   F  NaN   
4         New Smyrna Beach, Volusia County          Surfing   F   22   

                                              injury fatal   time  \
0                      Minor lacerations to left leg     N  11h00   
1      Lacerations to right calf and posterior thigh     N  09h30   
2  Injury to left forearm by hooked shark PROVOKE...     N  16h00   
3                      Lacerations to hand and wrist     N  14h00   
4                               Lacerations to foot      N  17h00   

                species  
0                   NaN  
1  White shark, 2-to 3m  
2    Blacktip shark, 6'  
3       Nurse shark, 5'  
4                   NaN

In [11]:
# Check information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25792 entries, 0 to 25791
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   date      6558 non-null   object 
 1   year      6556 non-null   float64
 2   type      6552 non-null   object 
 3   country   6508 non-null   object 
 4   area      6091 non-null   object 
 5   location  6008 non-null   object 
 6   activity  6002 non-null   object 
 7   sex       5987 non-null   object 
 8   age       3660 non-null   object 
 9   injury    6528 non-null   object 
 10  fatal     6006 non-null   object 
 11  time      3139 non-null   object 
 12  species   3610 non-null   object 
dtypes: float64(1), object(12)
memory usage: 2.6+ MB


## Rows

### Rows containing only missing values

There is total of 25792 lines, but all columns do not have even 9,000 rows with `non-NaN`. So,as mentioned before, there are a lot of rows that only contains missing values `NaN` and they can be removed. 

In [12]:
# Remove all rows that contains only NaN
df = df.dropna(how='all')

# Check the result
print(f'Number of rows after removing the rows containing only missing values: {df.shape[0]}\n')

# Check information about the dataframe
df.info()

Number of rows after removing the rows containing only missing values: 6558

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6558 entries, 0 to 6557
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   date      6558 non-null   object 
 1   year      6556 non-null   float64
 2   type      6552 non-null   object 
 3   country   6508 non-null   object 
 4   area      6091 non-null   object 
 5   location  6008 non-null   object 
 6   activity  6002 non-null   object 
 7   sex       5987 non-null   object 
 8   age       3660 non-null   object 
 9   injury    6528 non-null   object 
 10  fatal     6006 non-null   object 
 11  time      3139 non-null   object 
 12  species   3610 non-null   object 
dtypes: float64(1), object(12)
memory usage: 717.3+ KB


### Duplicate rows

In [13]:
# Check if there are duplicate rows
df.duplicated().sum()

7

In [14]:
# Check duplicated rows
df[df.duplicated(keep=False)]

date    year        type         country  \
3577  Reported 26-Jun-1972  1972.0  Unprovoked       AUSTRALIA   
3578  Reported 26-Jun-1972  1972.0  Unprovoked       AUSTRALIA   
3622                  1971  1971.0  Unprovoked            IRAN   
3623                  1971  1971.0  Unprovoked            IRAN   
4490              Aug-1956  1956.0    Provoked  UNITED KINGDOM   
4491              Aug-1956  1956.0    Provoked  UNITED KINGDOM   
4940             Fall 1943  1943.0  Unprovoked             USA   
4941             Fall 1943  1943.0  Unprovoked             USA   
5719  Reported 10-Oct-1906  1906.0  Unprovoked             USA   
5721  Reported 10-Oct-1906  1906.0  Unprovoked             USA   
5962                  1890  1890.0  Unprovoked           INDIA   
5963                  1890  1890.0  Unprovoked           INDIA   
6215                  1863  1863.0  Unprovoked          CEYLON   
6216                  1863  1863.0  Unprovoked          CEYLON   

                    area                                      location  \
3577          Queensland                                 Pancake Creek   
3578          Queensland                                 Pancake Creek   
3622  Khuzestan Province                     Ahvaz, on the Karun River   
3623  Khuzestan Province                     Ahvaz, on the Karun River   
4490            Cornwall                                    The Lizard   
4491            Cornwall                                    The Lizard   
4940              Hawaii  Midway Island, Northwestern Hawaiian Islands   
4941              Hawaii  Midway Island, Northwestern Hawaiian Islands   
5719              Hawaii                                           NaN   
5721              Hawaii                                           NaN   
5962          Tamil Nadu                                     Tuticorin   
5963          Tamil Nadu                                     Tuticorin   
6215                 NaN                                           NaN   
6216                 NaN                                           NaN   

                                        activity sex  age  \
3577                                         NaN   M  NaN   
3578                                         NaN   M  NaN   
3622                                         NaN   M  NaN   
3623                                         NaN   M  NaN   
4490  Attempting to kill a shark with explosives   M  NaN   
4491  Attempting to kill a shark with explosives   M  NaN   
4940                                Spearfishing   M  NaN   
4941                                Spearfishing   M  NaN   
5719                                    Swimming   M  NaN   
5721                                    Swimming   M  NaN   
5962                                      Diving   M  NaN   
5963                                      Diving   M  NaN   
6215                              Fell overboard   M  NaN   
6216                              Fell overboard   M  NaN   

                        injury    fatal time         species  
3577                     FATAL        Y  NaN             NaN  
3578                     FATAL        Y  NaN             NaN  
3622                  Survived        N  NaN             NaN  
3623                  Survived        N  NaN             NaN  
4490  FATAL, PROVOKED INCIDENT        Y  NaN             NaN  
4491  FATAL, PROVOKED INCIDENT        Y  NaN             NaN  
4940  Calf nipped in each case        N  NaN  "small sharks"  
4941  Calf nipped in each case        N  NaN  "small sharks"  
5719                     FATAL        Y  NaN             NaN  
5721                     FATAL        Y  NaN             NaN  
5962                No details  UNKNOWN  NaN             NaN  
5963                No details  UNKNOWN  NaN             NaN  
6215                     FATAL        Y  NaN             NaN  
6216                     FATAL        Y  NaN             NaN

As shown above, there are some rows that are duplicated. Therefore, they can be removed.

In [15]:
# Remove duplicated rows
df = df.drop_duplicates(keep='first')

# Check the result
df.duplicated().sum()

0

In [16]:
# Check the new number of rows
n_rows = df.shape[0]
n_rows_initial = df.shape[0]
print(f'After removing the duplicates, the dataframe has {n_rows} rows.')

After removing the duplicates, the dataframe has 6551 rows.


# Specific Data Cleaning & Manipulation - Columns

## Gender

The gender is represented by the column `sex`.

In [17]:
# Check the unique values in the column
print(df.sex.unique())

['F' 'M' nan 'M ' 'lli' 'M x 2' 'N' '.']


In [18]:
# Check the number of each value
df.sex.value_counts(dropna=False)

M        5280
F         693
NaN       571
M           2
N           2
M x 2       1
lli         1
.           1
Name: sex, dtype: int64

In [19]:
# Remove unnecessary spaces in the values
df.sex = df.sex.str.strip()

# Check the result
df.sex.value_counts(dropna=False) / n_rows

M        0.806289
F        0.105785
NaN      0.087162
N        0.000305
M x 2    0.000153
lli      0.000153
.        0.000153
Name: sex, dtype: float64

In [20]:
# Check the rows with values that are not 'M', nor 'F', nor NaN
print(f'They represent only {(len(df[(df.sex != "M") & (df.sex != "F") & (df.sex == df.sex)].index) / n_rows) * 100:.2f}',
      f'% of the current dataframe.', sep='')
df[(df.sex != 'M') & (df.sex != 'F') & (df.sex == df.sex)]

They represent only 0.08% of the current dataframe.


date    year          type           country  \
1867           11-Nov-2004  2004.0    Unprovoked               USA   
3205           23-Oct-1962  1982.0  Sea Disaster               USA   
5191           11-Jul-1934  1934.0    Watercraft         AUSTRALIA   
5690  Reported 02-Jun-1908  1908.0  Sea Disaster  PAPUA NEW GUINEA   
6386  Reported 18-Dec-1801  1801.0      Provoked               NaN   

                 area                                        location  \
1867       California  Bunkers, Humboldt Bay, Eureka, Humboldt County   
3205   Carolina coast                                             NaN   
5191  New South Wales                                        Cronulla   
5690      New Britain                                          Matupi   
6386              NaN                                             NaN   

                              activity    sex  age  \
1867                           Surfing    lli   38   
3205  Yacht Trashman capsized in storm  M x 2  NaN   
5191                           Fishing      N  NaN   
5690                                 .      .  NaN   
6386   Standing on landed shark's tail      N  NaN   

                                                 injury  fatal   time  \
1867                 Lacerations to hand, knee & thigh       N  13h30   
3205                                              FATAL  Y x 2    NaN   
5191  No injury to occupants Sharks continually foll...      N    NaN   
5690  Remains of 3 humans recovered from shark, but ...      Y    NaN   
6386                           FATAL, PROVOKED INCIDENT      Y    NaN   

                        species  
1867    5.5 m [18'] white shark  
3205                        NaN  
5191         Blue pointer, 11'   
5690  Allegedly a 33-foot shark  
6386                  12' shark

Given the low percentage, they will not be considered in the analysis, in other other, they will be removed.

In [21]:
print(f'Before the removal, there was {n_rows} rows.', end=' ')

# Remove the rows mentioned above
df = df.drop(df[(df.sex != 'M') & (df.sex != 'F') & (df.sex == df.sex)].index)

# Number of rows
n_rows = df.shape[0]
print(f'After the removal, there is {n_rows} rows.')

Before the removal, there was 6551 rows. After the removal, there is 6546 rows.


In [22]:
# Check rows with missing values in the 'sex' column
df[df.sex.isna()]

date    year          type           country  \
8              29-Jul-2020  2020.0    Watercraft         AUSTRALIA   
22             20-Jun-2020  2020.0    Unprovoked           BAHAMAS   
32             17 May 2020  2020.0    Unprovoked  FRENCH POLYNESIA   
48             05-Feb-2020  2020.0    Unprovoked               USA   
65             20-Dec-2019  2019.0      Provoked         AUSTRALIA   
...                    ...     ...           ...               ...   
6468           Before 2004     0.0    Watercraft        MOZAMBIQUE   
6493  No date, Before 1963     0.0    Unprovoked         SINGAPORE   
6515             1941-1945     0.0  Sea Disaster               NaN   
6534   Between 1918 & 1939     0.0    Unprovoked           REUNION   
6543           Before 1917     0.0    Unprovoked              FIJI   

                    area                                           location  \
8               Tasmania                                       Tenth Island   
22                Exumas                              Pig Beach, Pig Island   
32                Tahiti                                     Vallée Blanche   
48                  Maui                                                NaN   
65       New South Wales                                       Shellharbour   
...                  ...                                                ...   
6468  Inhambane Province                                      Off Inhambane   
6493                 NaN  Keppel Harbor,  2 miles from Singapore city ce...   
6515                 NaN                                                NaN   
6534         Saint-Denis                                          Barachois   
6543        Moala Island                                                NaN   

                                        activity  sex  age  \
8                                    Sightseeing  NaN  NaN   
22                                           NaN  NaN  NaN   
32                                    Snorkeling  NaN  NaN   
48                      Stand-Up Paddle boarding  NaN  NaN   
65                                       Fishing  NaN  NaN   
...                                          ...  ...  ...   
6468                                     Fishing  NaN  NaN   
6493                                    Swimming  NaN  NaN   
6515  A group of survivors on a raft for 17-days  NaN  NaN   
6534                                    Swimming  NaN  NaN   
6543        Wreck of  large double sailing canoe  NaN  NaN   

                                                 injury fatal            time  \
8     No injury to occupants, injury to shark attemp...   NaN           09h08   
22                                                  NaN     N             NaN   
32                                           Leg bitten     N             NaN   
48                    No injury, but paddleboard bitten     N           09h40   
65                                    PROVOKED INCIDENT   NaN             NaN   
...                                                 ...   ...             ...   
6468          No injury to occupants, shark bumped boat     N             NaN   
6493                                          Recovered     N             NaN   
6515  FATAL, shark leapt into raft and bit the man w...     Y  Late afternoon   
6534                                              FATAL     Y             NaN   
6543  FATAL, 18 people  were killed by sharks, 2 sur...     Y             NaN   

               species  
8      White shark, 4m  
22                 NaN  
32      Blacktip shark  
48         Tiger shark  
65         White shark  
...                ...  
6468       Whale shark  
6493               NaN  
6515  1.2 m [4'] shark  
6534               NaN  
6543               NaN  

[571 rows x 13 columns]

In [23]:
# Check % of missing values
df.sex.isna().mean()*100

8.722884204094102

For now, they will be remain in the dataframe, but they may be removed later.

## Age

### Data Cleaning

In [24]:
# Check the unique values in the column
print(f'There are {len(list(df.age.unique()))} unique values in this column.\n')
print(df.age.unique())

There are 237 unique values in this column.

[50 35 55 nan 22 14 28 38 4 63 23 11 12 10 29 15 36 7 16 30 60 18 9 26 57
 'Teen' 24 59 13 75 21 '30s' 45 33 17 37 70 44 '28 & 22' 32 20 51
 '22, 57, 31' '60s' 40 49 "20's" 43 8 64 19 65 67 53 34 25 58 74 46 41 31
 '9 & 60' 48 '20s' 42 39 56 61 'a minor' 6 62 52 54 69 '40s' 3 82 73 68 47
 66 72 27 71 '38' '39' '23' '32' '52' '68' '12' '18' '19' '43' '47' '6'
 '37' '9' '36' '10' '16' '13' '11' '17' '14' '30' '50' '29' '65' '63' '26'
 '71' '48' '70' '58' '18 months' '22' '41' '35' '57' '20' '24' '34' '15'
 '44' '53' '7' '40' '28' '33' '31' '45' '50s' '8' '51' '61' '42' '25'
 'teen' '66' '21' '77' '46' '60' '74' '55' '27' '3' '56' '64' '28 & 26'
 '62' '5' '49' '54' '86' '59' '18 or 20' '12 or 13' '46 & 34'
 '28, 23 & 30' 'Teens' 77 '36 & 26' '8 or 10' 84 '\xa0 ' ' ' '30 or 36'
 '6½' '21 & ?' '33 or 37' 'mid-30s' '23 & 20' 5 ' 30' '7      &    31'
 ' 28' '20?' "60's" '69' '32 & 30' '16 to 18' '87' '67' 'Elderly'
 'mid-20s' 'Ca. 33' '74 ' '45 ' '

In [25]:
# Check the number of each value
df.age.value_counts(dropna=False)

NaN                        2887
16                           86
19                           86
17                           82
18                           82
                           ... 
17 & 35                       1
2 to 3 months                 1
X                             1
37, 67, 35, 27,  ? & 27       1
33 or 37                      1
Name: age, Length: 237, dtype: int64

In [26]:
# Convert all values in this columnn to string if it is not a missing value
df.age = df.age.apply(lambda x : str(x) if x == x else x)

# Check result
df.age.unique()

array(['50', '35', '55', nan, '22', '14', '28', '38', '4', '63', '23',
       '11', '12', '10', '29', '15', '36', '7', '16', '30', '60', '18',
       '9', '26', '57', 'Teen', '24', '59', '13', '75', '21', '30s', '45',
       '33', '17', '37', '70', '44', '28 & 22', '32', '20', '51',
       '22, 57, 31', '60s', '40', '49', "20's", '43', '8', '64', '19',
       '65', '67', '53', '34', '25', '58', '74', '46', '41', '31',
       '9 & 60', '48', '20s', '42', '39', '56', '61', 'a minor', '6',
       '62', '52', '54', '69', '40s', '3', '82', '73', '68', '47', '66',
       '72', '27', '71', '18 months', '50s', 'teen', '77', '28 & 26', '5',
       '86', '18 or 20', '12 or 13', '46 & 34', '28, 23 & 30', 'Teens',
       '36 & 26', '8 or 10', '84', '\xa0 ', ' ', '30 or 36', '6½',
       '21 & ?', '33 or 37', 'mid-30s', '23 & 20', ' 30',
       '7      &    31', ' 28', '20?', "60's", '32 & 30', '16 to 18',
       '87', 'Elderly', 'mid-20s', 'Ca. 33', '74 ', '45 ', '21 or 26',
       '20 ', '>50', '

Converting all the values to the same data type make it easier to manipulate them.

In [27]:
len(df.age.unique())

164

In [28]:
# Check rows that contains '&'
df[df.age.str.contains('&', na=False)]

date    year          type             country  \
80             29-Oct-2019  2019.0    Unprovoked           AUSTRALIA   
168            10-Jan-2019  2019.0       Invalid           AUSTRALIA   
940            10-Mar-2013  2013.0    Unprovoked         PHILIPPINES   
1749  Reported 28-Jan-2006  2006.0    Watercraft      ATLANTIC OCEAN   
1751           23-Jan-2006  2006.0    Watercraft      ATLANTIC OCEAN   
1986           14-Sep-2003  2003.0    Watercraft        SOUTH AFRICA   
2168           12-Aug-2001  2001.0    Unprovoked            THAILAND   
2269           10-Aug-2000  2000.0       Invalid                 USA   
2665  Reported 11-Sep-1994  1994.0  Sea Disaster                 USA   
2754           04-Jan-1993  1993.0    Watercraft       CARIBBEAN SEA   
3709           24-Aug-1968  1968.0    Unprovoked                 USA   
3767           10-Aug-1967  1967.0       Invalid           AUSTRALIA   
3992              Aug-1963  1963.0    Unprovoked  DOMINICAN REPUBLIC   
4130  Reported 06-Sep-1961  1961.0      Provoked               ITALY   
4236           07-Jun-1960  1960.0       Invalid                 USA   
4248           14-Apr-1960  1960.0       Invalid             BERMUDA   
4259           29-Feb-1960  1960.0    Unprovoked           AUSTRALIA   
4285           28-Dec-1959  1959.0    Watercraft           AUSTRALIA   
4289           11-Dec-1959  1959.0      Provoked           AUSTRALIA   
4328                 21764  1959.0       Invalid               ITALY   
4368           02-Jan-1959  1959.0    Unprovoked          MOZAMBIQUE   
4840           14-Jan-1947  1947.0    Watercraft           AUSTRALIA   
5252           06-Jan-1932  1932.0      Provoked              MEXICO   
5434           04-Jul-1924  1924.0    Watercraft                 USA   

                        area  \
80                Queensland   
168               Queensland   
940                  Palawan   
1749  300 miles from Antigua   
1751     800 miles from land   
1986   Western Cape Province   
2168         Rayong Province   
2269                 Florida   
2665                 Florida   
2754                     NaN   
3709                 Florida   
3767         New South Wales   
3992           Santo Domingo   
4130         Venice Province   
4236              California   
4248                33N, 68W   
4259                Tasmania   
4285         New South Wales   
4289                Victoria   
4328                 Tuscany   
4368         Maputo Province   
4840         New South Wales   
5252         Baja California   
5434              California   

                                               location  \
80                        Off Airlie Beach, Whitsundays   
168         Catseye Beach, Hamilton Island, Whitsundays   
940                                    Off Likas Island   
1749                                                NaN   
1751                                                NaN   
1986                                      Melkbosstrand   
2168                                 Laem Mae Pim Beach   
2269                                       Florida Keys   
2665                                    Florida Straits   
2754                             Off Dominican Republic   
3709                 Half mile north of Juno Beach Pier   
3767                                         Jervis Bay   
3992                                     Rio Haina Port   
4130                                           Chioggia   
4236   10 miles off Santa Barbara, Santa Barbara County   
4248  Royal Canadian Navy CS2F-1 aircraft ditched in...   
4259                                          Ralph Bay   
4285                                  Leichardt, Sydney   
4289                                  Altona, Melbourne   
4328                   Cala del Corvo, Isola del Giglio   
4368                                  Off Inhaca Island   
4840                        Berry's Bay, Sydney Harbour   
5252                                      Descano Point   
5434

In [29]:
# Number of rows that contain "&"
r_clean = len(df[df.age.str.contains("&", na=False)])
print(f'There are {r_clean} rows containing "&" and they represent only {(r_clean / n_rows) * 100:.2f}% of the', end=' '
      f'current dataframe.\nSince they have ages of more than one person, they will be removed from the dataframe.', sep=' ')

There are 24 rows containing "&" and they represent only 0.37% of the current dataframe.
Since they have ages of more than one person, they will be removed from the dataframe.

In [30]:
print(f'Before the removal, there was {n_rows} rows.', end=' ')

# Remove the rows mentioned above
df = df.drop(df[df.age.str.contains('&', na=False)].index)

# Number of rows
n_rows = df.shape[0]
print(f'After the removal, there is {n_rows} rows.')

Before the removal, there was 6546 rows. After the removal, there is 6522 rows.


In [31]:
# Check the unique values
print(f'Now, there are {len(list(df.age.unique()))} unique values in this column.\n')
print(df.age.unique())

Now, there are 140 unique values in this column.

['50' '35' '55' nan '22' '14' '28' '38' '4' '63' '23' '11' '12' '10' '29'
 '15' '36' '7' '16' '30' '60' '18' '9' '26' '57' 'Teen' '24' '59' '13'
 '75' '21' '30s' '45' '33' '17' '37' '70' '44' '32' '20' '51' '22, 57, 31'
 '60s' '40' '49' "20's" '43' '8' '64' '19' '65' '67' '53' '34' '25' '58'
 '74' '46' '41' '31' '48' '20s' '42' '39' '56' '61' 'a minor' '6' '62'
 '52' '54' '69' '40s' '3' '82' '73' '68' '47' '66' '72' '27' '71'
 '18 months' '50s' 'teen' '77' '5' '86' '18 or 20' '12 or 13' 'Teens'
 '8 or 10' '84' '\xa0 ' ' ' '30 or 36' '6½' '33 or 37' 'mid-30s' ' 30'
 ' 28' '20?' "60's" '16 to 18' '87' 'Elderly' 'mid-20s' 'Ca. 33' '74 '
 '45 ' '21 or 26' '20 ' '>50' '18 to 22' 'adult' '9 months' '25 to 35' '1'
 '(adult)' '25 or 28' 'X' '"middle-age"' '13 or 18' '2 to 3 months'
 'MAKE LINE GREEN' ' 43' '81' '"young"' '7 or 8' '78' 'F' 'Both 11'
 '9 or 10' 'young' '  ' 'A.M.' '10 or 12' '31 or 33' '2½' '13 or 14']


In [32]:
# Check rows that the age values does not make sense or can not be categorized
df[df.age.apply(lambda x : True if x in ['\xa0 ', 'X', ' ', '20?', '>50', '13 or 18', 'MAKE LINE GREEN', 'Ca. 33',
                                         '"young"', 'F', 'Both 11', 'young', '  ', 'A.M.'] else False)]

date    year          type           country  \
2067  27-Sep-2002  2002.0      Provoked               USA   
2068  27-Sep-2002  2002.0      Provoked             TONGA   
2684  06-Apr-1994  1994.0    Unprovoked           BAHAMAS   
3229     May 1982  1982.0       Invalid               USA   
3355  29-Dec-1978  1978.0    Unprovoked         AUSTRALIA   
3612  11-Apr-1971  1971.0    Unprovoked      SOUTH AFRICA   
4302  07-Oct-1959  1959.0    Unprovoked        MOZAMBIQUE   
4325  11-Aug-1959  1959.0    Unprovoked             JAPAN   
4643  08-Jan-1953  1953.0    Watercraft         AUSTRALIA   
4700  24-Mar-1951  1951.0    Unprovoked         AUSTRALIA   
4929  14-May-1943  1943.0  Sea Disaster         AUSTRALIA   
5054  12-Jan-1939  1939.0    Unprovoked         AUSTRALIA   
5139  23-Feb-1936  1936.0      Provoked         AUSTRALIA   
5337  27-Jan-1929  1929.0    Unprovoked         AUSTRALIA   
5423  10-Mar-1925  1925.0    Unprovoked  PAPUA NEW GUINEA   
6398         1779  1779.0    Unprovoked               USA   

                       area                                  location  \
2067                Florida                  Key Largo, Monroe County   
2068                 Vava'u             Swimming with humpback whales   
2684                    NaN                                   Rum Cay   
3229         South Carolina  Daws Island, Broad River (near Beaufort)   
3355             Queensland                             Bribie Island   
3612  Western Cape Province                               Buffels Bay   
4302        Maputo Province              Xefina Island, Bay of Maputo   
4325    Wakayama Prefecture            Isonoura Beach, Wakayama City    
4643               Tasmania                                   Wynyard   
4700        New South Wales                                    Sydney   
4929             Queensland                              Off Brisbane   
5054        New South Wales                            Clarence River   
5139        New South Wales                      Angourie, near Yamba   
5337             Queensland     Alma Bay, Magnetic Island, Townsville   
5423       Central Province                                      Hula   
6398                 Hawaii                            Maliu, Hawai'i   

                                               activity  sex              age  \
2067                                            Fishing    M                    
2068                                           Swimming    M                    
2684                                             Wading    M              20?   
3229                                                NaN    F           Ca. 33   
3355                                                NaN    M                    
3612                                           Swimming    M              >50   
4302                                           Swimming    M                X   
4325                                           Swimming    M         13 or 18   
4643                                            Fishing  NaN  MAKE LINE GREEN   
4700                       Fishing, casting in the surf    M          "young"   
4929  Hospital Ship Centaur torpedoed & sunk by the ...    M                F   
5054                                    Scooping prawns    M          Both 11   
5139      Touching the mouth of a supposedly dead shark    F            young   
5337                                           Swimming    M                    
5423                                           Swimming    M             A.M.   
6398                                            Surfing    M            young   

                                                 injury fatal         time  \
2067             Left thumb lacerated PROVOKED INCIDENT     N    Afternoon   
2068                 Thigh lacerated  PROVOKED INCIDENT     N        15h00   
2684                            Lower right  leg bitten     N        10h30   
3229           Human remains recovered from

In [33]:
# Number of rows that have invalid age
r_clean = len(df[df.age.apply(lambda x : True if x in ['\xa0 ', 'X', ' ', '20?', '>50', '13 or 18', 'MAKE LINE GREEN',
                                                       'Ca. 33', '"young"', 'F', 'Both 11', 'young', '  ', 'A.M.'] 
                              else False)].index)
print(f'There are {r_clean} rows with invalid age and they represent only {(r_clean / n_rows) * 100:.2f}% of the', end=' '
      f'current dataframe.\nSince they can not be categorized, they will be removed from the dataframe', sep=' ')

There are 16 rows containing "&" and they represent only 0.25% of the current dataframe.
Since they can not be categorized, they will be removed from the dataframe

In [34]:
print(f'Before the removal, there was {n_rows} rows.', end=' ')

# Remove the rows mentioned above
df = df.drop(df[df.age.apply(lambda x : True if x in ['\xa0 ', 'X', ' ', '20?', '>50', '13 or 18', 'MAKE LINE GREEN', 
                                                      'Ca. 33', '"young"', 'F', 'Both 11', 'young', '  ', 'A.M.', '30s',
                                                     'a minor'] 
                             else False)].index)

# Number of rows
n_rows = df.shape[0]
print(f'After the removal, there is {n_rows} rows.')

Before the removal, there was 6522 rows. After the removal, there is 6497 rows.


In [35]:
# Check the unique values
print(f'Now, there are {len(list(df.age.unique()))} unique values in this column.\n')
print(df.age.unique())

Now, there are 124 unique values in this column.

['50' '35' '55' nan '22' '14' '28' '38' '4' '63' '23' '11' '12' '10' '29'
 '15' '36' '7' '16' '30' '60' '18' '9' '26' '57' 'Teen' '24' '59' '13'
 '75' '21' '45' '33' '17' '37' '70' '44' '32' '20' '51' '22, 57, 31' '60s'
 '40' '49' "20's" '43' '8' '64' '19' '65' '67' '53' '34' '25' '58' '74'
 '46' '41' '31' '48' '20s' '42' '39' '56' '61' '6' '62' '52' '54' '69'
 '40s' '3' '82' '73' '68' '47' '66' '72' '27' '71' '18 months' '50s'
 'teen' '77' '5' '86' '18 or 20' '12 or 13' 'Teens' '8 or 10' '84'
 '30 or 36' '6½' '33 or 37' 'mid-30s' ' 30' ' 28' "60's" '16 to 18' '87'
 'Elderly' 'mid-20s' '74 ' '45 ' '21 or 26' '20 ' '18 to 22' 'adult'
 '9 months' '25 to 35' '1' '(adult)' '25 or 28' '"middle-age"'
 '2 to 3 months' ' 43' '81' '7 or 8' '78' '9 or 10' '10 or 12' '31 or 33'
 '2½' '13 or 14']


In [36]:
# Check % of missing values
df.age.isna().mean()*100

44.43589348930275

The percentage of missing values is too high and, if they are removed, it would greatly impact the dataset. So, for now, they will remain in the dataframe and they will be analyzed again later.

### Categorizing the age

The column age will be categorized into:

<table>
    <thead>
        <tr>
            <th>AGE</th>
            <th>CATEGORY</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>< 11</td>
            <td>Child</td>
        </tr>
        <tr>
            <td>12 - 17</td>
            <td>Teenager</td>
        </tr>
        <tr>
            <td>18 - 35</td>
            <td>Young Adult</td>
        </tr>
        <tr>
            <td>36 - 65</td>
            <td>Adult</td>
        </tr>
        <tr>
            <td>> 65</td>
            <td>Elder</td>
        </tr>
        <tr>
            <td>NaN</td>
            <td>No category</td>
        </tr>
    </tbody>
</table>

In [37]:
# Create a function to categorize the ages
def cat_ages(age: str):
    '''
    Categorizes an age.
    
    Args:
    -----
        age (str): a string containing an age
    
    Returns:
    --------
        category (str): a string containing the category
    '''
    
    # Local variable that will store the category
    category = ''
    
    # If the age is a digit
    try:
        age = int(age)
        if age > 65:
            category = 'Elder'
        elif age > 35:
            category = 'Adult'
        elif age > 17:
            category = 'Young Adult'
        elif age > 11:
            category = 'Teenager'
        else:
            category = 'Child'
        
        return category
    
    # If the age is a missing value or not a digit
    except:
        if age != age:
            category = 'No category'
        elif age in ['18 months', '8 or 10', '6½', '9 months', '2 to 3 months', '7 or 8', '9 or 10', '10 or 12', '2½']:
            category = 'Child'
        elif age in ['Teen', 'teen', '12 or 13', 'Teens', '16 to 18', '13 or 14']:
            category = 'Teenager'
        elif age in ["20's", '20s', '18 or 20', 'mid-20s', '30 or 36', '21 or 26', '18 to 22', '25 to 35', '25 or 28',
                     '31 or 33']:
            category = 'Young Adult'
        elif age in ['40s', '50s', 'mid-30s', '33 or 37', 'adult', '(adult)', '"middle-age"']:
            category = 'Adult'
        elif age in ['60s', "60's", 'Elderly']:
            category = 'Elder'
        
        return category

In [38]:
# Categorize the ages
df['cat_age'] = df.age.apply(cat_ages)

# Check the result
df.cat_age.value_counts(dropna=False)

No category    2887
Young Adult    1723
Adult           840
Teenager        751
Child           235
Elder            60
                  1
Name: cat_age, dtype: int64

In [39]:
# Check number of rows
df.cat_age.value_counts(dropna=False).sum() == n_rows

True

In [40]:
# Check information about the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6497 entries, 0 to 6557
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   date      6497 non-null   object 
 1   year      6495 non-null   float64
 2   type      6491 non-null   object 
 3   country   6448 non-null   object 
 4   area      6034 non-null   object 
 5   location  5953 non-null   object 
 6   activity  5945 non-null   object 
 7   sex       5931 non-null   object 
 8   age       3610 non-null   object 
 9   injury    6467 non-null   object 
 10  fatal     5951 non-null   object 
 11  time      3115 non-null   object 
 12  species   3571 non-null   object 
 13  cat_age   6497 non-null   object 
dtypes: float64(1), object(13)
memory usage: 761.4+ KB


In [41]:
# Check the dataset
df.head()

date    year        type    country             area  \
0  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2  10-Aug-2020  2020.0    Provoked        USA          Florida   
3  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity sex  age  \
0         New Smyrna Beach, Volusia County  Boogie boarding   F   50   
1             Shelly Beach, Port Macquarie          Surfing   F   35   
2  Off Gasparilla Island, Charlotte County          Fishing   M   55   
3                   Candle Reef, St. Croix       Snorkeling   F  NaN   
4         New Smyrna Beach, Volusia County          Surfing   F   22   

                                              injury fatal   time  \
0                      Minor lacerations to left leg     N  11h00   
1      Lacerations to right calf and posterior thigh     N  09h30   
2  Injury to left forearm by hooked shark PROVOKE...     N  16h00   
3                      Lacerations to hand and wrist     N  14h00   
4                               Lacerations to foot      N  17h00   

                species      cat_age  
0                   NaN        Adult  
1  White shark, 2-to 3m  Young Adult  
2    Blacktip shark, 6'        Adult  
3       Nurse shark, 5'  No category  
4                   NaN  Young Adult

## Activity

In [42]:
# Check the unique values in the column
print(f'There are {len(list(df.activity.unique()))} unique values in this column.\n')
df.activity.unique()

There are 1531 unique values in this column.



array(['Boogie boarding', 'Surfing', 'Fishing', ...,
       'Crew swimming alongside their anchored ship',
       '4 men were bathing', 'Wreck of  large double sailing canoe'],
      dtype=object)

In [43]:
df.activity.unique()

array(['Boogie boarding', 'Surfing', 'Fishing', ...,
       'Crew swimming alongside their anchored ship',
       '4 men were bathing', 'Wreck of  large double sailing canoe'],
      dtype=object)

In [44]:
# Check the number of each value
df.activity.value_counts(dropna=False)

Surfing                                     1034
Swimming                                     888
NaN                                          552
Fishing                                      454
Spearfishing                                 354
                                            ... 
Checking fish traps, fell into the water       1
Jumping in swells                              1
Swimming to shore from capsized kayak          1
Free diving Spearfishing                       1
Wakeboarding                                   1
Name: activity, Length: 1531, dtype: int64

In [45]:
# Remove unnecessary spaces in the values
df.activity = df.activity.str.strip()

# Check the result
df.activity.value_counts(dropna=False)

Surfing                                                  1039
Swimming                                                  935
NaN                                                       552
Fishing                                                   469
Spearfishing                                              362
                                                         ... 
Free diving, working on U/W scenes for motion picture       1
Checking fish traps, fell into the water                    1
Jumping in swells                                           1
Swimming to shore from capsized kayak                       1
Wakeboarding                                                1
Name: activity, Length: 1502, dtype: int64

In [46]:
# Check the unique values in the column
print(f'Now, there are {len(list(df.activity.unique()))} unique values in this column.\n')
df.activity.unique()

Now, there are 1502 unique values in this column.



array(['Boogie boarding', 'Surfing', 'Fishing', ...,
       'Crew swimming alongside their anchored ship',
       '4 men were bathing', 'Wreck of  large double sailing canoe'],
      dtype=object)

For this analysis, only the activity that was involved the most in the shark incidents is important, but I will check the top 10 activities.

In [47]:
# Check the top 10 actvities
df.activity.value_counts(dropna=False).head(10)

Surfing         1039
Swimming         935
NaN              552
Fishing          469
Spearfishing     362
Bathing          166
Wading           160
Diving           134
Standing         106
Snorkeling       102
Name: activity, dtype: int64

## Year

### Data Cleaning

In [48]:
# Check infromation about the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6497 entries, 0 to 6557
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   date      6497 non-null   object 
 1   year      6495 non-null   float64
 2   type      6491 non-null   object 
 3   country   6448 non-null   object 
 4   area      6034 non-null   object 
 5   location  5953 non-null   object 
 6   activity  5945 non-null   object 
 7   sex       5931 non-null   object 
 8   age       3610 non-null   object 
 9   injury    6467 non-null   object 
 10  fatal     5951 non-null   object 
 11  time      3115 non-null   object 
 12  species   3571 non-null   object 
 13  cat_age   6497 non-null   object 
dtypes: float64(1), object(13)
memory usage: 761.4+ KB


It is clear that the column `year` has only 2 missing values.

In [49]:
# Check values in the column
print(f'There are {len(list(df.year.unique()))} unique values in this column.\n')
df.year.unique()

There are 250 unique values in this column.



array([2020., 2019., 2018., 2017.,   nan, 2016., 2015., 2014., 2013.,
       2012., 2011., 2010., 2009., 2008., 2007., 2006., 2005., 2004.,
       2003., 2002., 2001., 2000., 1999., 1998., 1997., 1996., 1995.,
       1984., 1994., 1993., 1992., 1991., 1990., 1989., 1969., 1988.,
       1987., 1986., 1985., 1983., 1982., 1981., 1980., 1979., 1978.,
       1977., 1976., 1975., 1974., 1973., 1972., 1971., 1970., 1968.,
       1967., 1966., 1965., 1964., 1963., 1962., 1961., 1960., 1959.,
       1958., 1957., 1956., 1955., 1954., 1953., 1952., 1951., 1950.,
       1949., 1948., 1848., 1947., 1946., 1945., 1944., 1943., 1942.,
       1941., 1940., 1939., 1938., 1937., 1936., 1935., 1934., 1933.,
       1932., 1931., 1930., 1929., 1928., 1927., 1926., 1925., 1924.,
       1923., 1922., 1921., 1920., 1919., 1918., 1917., 1916., 1915.,
       1914., 1913., 1912., 1911., 1910., 1909., 1908., 1907., 1906.,
       1905., 1904., 1903., 1902., 1901., 1900., 1899., 1898., 1897.,
       1896., 1895.,

The idea for the column `year` is to categorize it to century from XIX to XXI.

In [50]:
# Check rows that the value is before 1801 or a missing value
date_before_xix = df[(df.year < 1801) | (df.year.isna())]
date_before_xix

date    year        type             country  \
428   Reported 08-Jan-2017     NaN     Invalid           AUSTRALIA   
6334  Reported 19-Aug-1836     NaN  Unprovoked             ENGLAND   
6387                  1800  1800.0  Unprovoked          SEYCHELLES   
6388  Reported May-28-1797  1797.0  Unprovoked                 NaN   
6389           12-Sep-1792  1792.0    Provoked             ENGLAND   
...                    ...     ...         ...                 ...   
6553           Before 1903     0.0  Unprovoked           AUSTRALIA   
6554           Before 1903     0.0  Unprovoked           AUSTRALIA   
6555             1900-1905     0.0  Unprovoked                 USA   
6556             1883-1889     0.0  Unprovoked              PANAMA   
6557             1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                   area                             location  \
428          Queensland                                  NaN   
6334         Cumberland                           Whitehaven   
6387           St. Anne                                  NaN   
6388                NaN                                  NaN   
6389                NaN                  On the Thames River   
...                 ...                                  ...   
6553  Western Australia                          Roebuck Bay   
6554  Western Australia                                  NaN   
6555     North Carolina                       Ocracoke Inlet   
6556                NaN                 Panama Bay 8ºN, 79ºW   
6557   Eastern Province  Below the English fort, Trincomalee   

                             activity  sex  age  \
428                      Spearfishing    M   35   
6334                         Swimming    M  NaN   
6387  a corsair's boat was overturned    F  NaN   
6388                Dropped overboard  NaN  NaN   
6389                          Fishing    M  NaN   
...                               ...  ...  ...   
6553                           Diving    M  NaN   
6554                     Pearl diving    M  NaN   
6555                         Swimming    M  NaN   
6556                              NaN    M  NaN   
6557                         Swimming    M   15   

                                                 injury fatal time  \
428              No attack, shark made a threat display   NaN  NaN   
6334                                              FATAL     Y  NaN   
6387           FATAL, all onboard were killed by sharks     Y  NaN   
6388                                              FATAL     Y  NaN   
6389                           FATAL, PROVOKED INCIDENT     Y  NaN   
...                                                 ...   ...  ...   
6553                                              FATAL     Y  NaN   
6554                                              FATAL     Y  NaN   
6555                                              FATAL     Y  NaN   
6556                                              FATAL     Y  NaN   
6557  FATAL. "Shark bit him in half, carrying away t...     Y  NaN   

         species      cat_age  
428   Bull shark  Young Adult  
6334         NaN  No category  
6387         NaN  No category  
6388         NaN  No category  
6389         NaN  No category  
...          ...          ...  
6553         NaN  No category  
6554         NaN  No category  
6555         NaN  No category  
6556         NaN  No category  
6557         NaN     Teenager  

[172 rows x 14 columns]

In [51]:
print(f'There are {len(date_before_xix.index)} rows that have this condition and they represent only '
      f'{len(date_before_xix.index) / n_rows:.2f}% of the current dataframe, they will be removed.', sep=' ')

There are 172 rows that have this condition and they represent only 0.03% of the current dataframe, they will be removed.


In [52]:
print(f'Before the removal, there was {n_rows} rows.', end=' ')

# Remove the rows mentioned above
df = df.drop(date_before_xix.index)

# Number of rows
n_rows = df.shape[0]
print(f'After the removal, there is {n_rows} rows.')

Before the removal, there was 6497 rows. After the removal, there is 6325 rows.


### Categorizing the year

The column year will be categorized into:

<table>
    <thead>
        <tr>
            <th>YEARS</th>
            <th>CATEGORY</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>1801 - 1900</td>
            <td>XIX</td>
        </tr>
        <tr>
            <td>1901 - 2000</td>
            <td>XX</td>
        </tr>
        <tr>
            <td>> 2000</td>
            <td>XIX</td>
        </tr>
    </tbody>
</table>

In [53]:
def cat_year(year):
    '''
    Categorizes a year to a century.
    
    Args:
    -----
        age (float): a floatg containing a year
    
    Returns:
    --------
        category (str): a string containing the century
    '''
    
    category = None
    
    if 1800 < year <= 1900:
        category = 'XIX'
    elif 1900 < year <= 2000:
        category = 'XX'
    else:
        category = 'XXI'
    
    return category

In [54]:
# Categorize the years
df['century'] = df.year.apply(cat_year)

# Check the result
df.century.value_counts(dropna=False)

XX     3538
XXI    2207
XIX     580
Name: century, dtype: int64

In [55]:
# Check number of rows
df.century.value_counts(dropna=False).sum() == n_rows

True

# Final Dataframe to use in the analysis

In [56]:
print(f'After cleaning the dataset, only {1 - (n_rows / n_rows_initial):.2f}% was removed.')

After cleaning the dataset, only 0.03% was removed.


In [57]:
df.head()

date    year        type    country             area  \
0  20-Aug-2020  2020.0  Unprovoked        USA          Florida   
1  14-Aug-2020  2020.0  Unprovoked  AUSTRALIA  New South Wales   
2  10-Aug-2020  2020.0    Provoked        USA          Florida   
3  02-Aug-2020  2020.0  Unprovoked        USA   Virgin Islands   
4  31-Jul-2020  2020.0  Unprovoked        USA          Florida   

                                  location         activity sex  age  \
0         New Smyrna Beach, Volusia County  Boogie boarding   F   50   
1             Shelly Beach, Port Macquarie          Surfing   F   35   
2  Off Gasparilla Island, Charlotte County          Fishing   M   55   
3                   Candle Reef, St. Croix       Snorkeling   F  NaN   
4         New Smyrna Beach, Volusia County          Surfing   F   22   

                                              injury fatal   time  \
0                      Minor lacerations to left leg     N  11h00   
1      Lacerations to right calf and posterior thigh     N  09h30   
2  Injury to left forearm by hooked shark PROVOKE...     N  16h00   
3                      Lacerations to hand and wrist     N  14h00   
4                               Lacerations to foot      N  17h00   

                species      cat_age century  
0                   NaN        Adult     XXI  
1  White shark, 2-to 3m  Young Adult     XXI  
2    Blacktip shark, 6'        Adult     XXI  
3       Nurse shark, 5'  No category     XXI  
4                   NaN  Young Adult     XXI

In [58]:
# Select only the necessary columns
df = df.loc[:, ['year', 'activity', 'sex', 'age', 'cat_age', 'century']]

In [59]:
# Check information about the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6325 entries, 0 to 6385
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   year      6325 non-null   float64
 1   activity  5804 non-null   object 
 2   sex       5773 non-null   object 
 3   age       3594 non-null   object 
 4   cat_age   6325 non-null   object 
 5   century   6325 non-null   object 
dtypes: float64(1), object(5)
memory usage: 345.9+ KB


In [60]:
# Check the dataset
df.head()

year         activity sex  age      cat_age century
0  2020.0  Boogie boarding   F   50        Adult     XXI
1  2020.0          Surfing   F   35  Young Adult     XXI
2  2020.0          Fishing   M   55        Adult     XXI
3  2020.0       Snorkeling   F  NaN  No category     XXI
4  2020.0          Surfing   F   22  Young Adult     XXI

In [61]:
# Create a backup version
df_bkp = df.copy()

# Analyzing the dataset

As mentioned in the **Introduction Section**, the following characteristics will be considered in the analysis:

- Gender
- Age
- Activity

## Gender

In [62]:
# Number of each value
df.sex.value_counts(dropna=False)

M      5103
F       670
NaN     552
Name: sex, dtype: int64

In [63]:
# % of each value
(df.sex.value_counts(dropna=False) / n_rows) * 100

M      80.679842
F      10.592885
NaN     8.727273
Name: sex, dtype: float64

It is clearly that the predominant gender involved in shark incidents is the **male** one.

In [64]:
# Filter the dataframe
df_male = df[df.sex == 'M']
n_rows_male = df_male.shape[0]

In [65]:
# Check information about the new dataframe
df_male.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5103 entries, 2 to 6384
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   year      5103 non-null   float64
 1   activity  4782 non-null   object 
 2   sex       5103 non-null   object 
 3   age       3066 non-null   object 
 4   cat_age   5103 non-null   object 
 5   century   5103 non-null   object 
dtypes: float64(1), object(5)
memory usage: 279.1+ KB


In [66]:
# Check the newdataframe
df_male.head()

year  activity sex age      cat_age century
2  2020.0   Fishing   M  55        Adult     XXI
5  2020.0  Swimming   M  14     Teenager     XXI
6  2020.0   Surfing   M  28  Young Adult     XXI
7  2020.0    Diving   M  38        Adult     XXI
9  2020.0  Swimming   M   4        Child     XXI

## Age

In [67]:
# Number of each value
df_male.cat_age.value_counts(dropna=False)

No category    2037
Young Adult    1527
Adult           712
Teenager        616
Child           161
Elder            50
Name: cat_age, dtype: int64

In [68]:
# % of each value
(df_male.cat_age.value_counts(dropna=False) / n_rows_male) * 100

No category    39.917695
Young Adult    29.923574
Adult          13.952577
Teenager       12.071331
Child           3.155007
Elder           0.979816
Name: cat_age, dtype: float64

Almost 40% of the dataset could not be categorized regarding the age. However, the one that stood out with the second highest percentage is the category **young adult**.

In [69]:
# 'Remove' the rows with missing values in the column 'age'
df_male_age = df_male[~(df_male.age.isna())]
n_rows_male_age = df_male_age.shape[0]
df_male_age.cat_age.value_counts(dropna=False)

Young Adult    1527
Adult           712
Teenager        616
Child           161
Elder            50
Name: cat_age, dtype: int64

In [70]:
# % of each value
(df_male_age.cat_age.value_counts(dropna=False) / n_rows_male_age) * 100

Young Adult    49.804305
Adult          23.222440
Teenager       20.091324
Child           5.251142
Elder           1.630789
Name: cat_age, dtype: float64

Disregarding rows containing missing values in the column `age`, almost half of the incidents with **males** happened to **young adults**.

In [71]:
# Filter the dataframe
df_male_ya = df_male[df_male.cat_age == 'Young Adult']
n_rows_male_ya = df_male_ya.shape[0]

In [72]:
# Check information about the new dataframe
df_male_ya.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1527 entries, 6 to 6369
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   year      1527 non-null   float64
 1   activity  1474 non-null   object 
 2   sex       1527 non-null   object 
 3   age       1527 non-null   object 
 4   cat_age   1527 non-null   object 
 5   century   1527 non-null   object 
dtypes: float64(1), object(5)
memory usage: 83.5+ KB


In [73]:
# Check the newdataframe
df_male_ya.head()

year      activity sex age      cat_age century
6   2020.0       Surfing   M  28  Young Adult     XXI
11  2020.0  Spearfishing   M  23  Young Adult     XXI
23  2020.0   Windsurfing   M  30  Young Adult     XXI
29  2020.0  Spearfishing   M  18  Young Adult     XXI
33  2020.0       Surfing   M  26  Young Adult     XXI

## Activity

In [74]:
# Number of each value
df_male_ya.activity.value_counts(dropna=False)

Surfing                                                                       423
Swimming                                                                      207
Spearfishing                                                                  155
Fishing                                                                        55
NaN                                                                            53
                                                                             ... 
Scuba Diving for lobster (at surface)                                           1
Swimming or surfing                                                             1
Swimming at edge of channel                                                     1
Swimming / Body surfing                                                         1
Fishing, when line became snagged on rock & he dived into water to free it      1
Name: activity, Length: 354, dtype: int64

In [75]:
# % of each value
(df_male_ya.activity.value_counts(dropna=False) / n_rows_male_ya) * 100

Surfing                                                                       27.701375
Swimming                                                                      13.555992
Spearfishing                                                                  10.150622
Fishing                                                                        3.601834
NaN                                                                            3.470858
                                                                                ...    
Scuba Diving for lobster (at surface)                                          0.065488
Swimming or surfing                                                            0.065488
Swimming at edge of channel                                                    0.065488
Swimming / Body surfing                                                        0.065488
Fishing, when line became snagged on rock & he dived into water to free it     0.065488
Name: activity, Length: 354, dty

Among the **young adult males**, the activity that was the most involved in shark incidents was the **surfing**.

In [76]:
# Filter the dataframe
df_male_ya_surfing = df_male_ya[df_male_ya.activity == 'Surfing']
n_rows_male_ya_surfing = df_male_ya_surfing.shape[0]

# Check the result
df_male_ya_surfing

year activity sex age      cat_age century
6     2020.0  Surfing   M  28  Young Adult     XXI
33    2020.0  Surfing   M  26  Young Adult     XXI
53    2020.0  Surfing   M  26  Young Adult     XXI
75    2019.0  Surfing   M  30  Young Adult     XXI
81    2019.0  Surfing   M  32  Young Adult     XXI
...      ...      ...  ..  ..          ...     ...
3990  1963.0  Surfing   M  18  Young Adult      XX
4098  1962.0  Surfing   M  19  Young Adult      XX
4534  1955.0  Surfing   M  25  Young Adult      XX
4772  1949.0  Surfing   M  32  Young Adult      XX
5488  1922.0  Surfing   M  26  Young Adult      XX

[423 rows x 6 columns]

# Conclusion

The most common characteristics of people involved in shark incidents in the history are:

- **Male**
- **Young adult**
- **Surfing**

# Exporting dataset

The clean dataframe used for the analysis will be export into a `.csv` file.

In [77]:
df.to_csv('sharks_clean.csv')

# Extra - Analysis though centuries

## Setup

In [78]:
# Create subsets for each century
df_cen21 = df.loc[df['century'] == 'XXI', :]
df_cen20 = df.loc[df['century'] == 'XX', :]
df_cen19 = df.loc[df['century'] == 'XIX', :]

In [79]:
# Check the dataset for the 'XIX' century
n_rows_cen19 = df_cen19.shape[0]
print(f'There are {n_rows_cen19} cases that happened during the XIX century.')
df_cen19.info()

There are 580 cases that happened during the XIX century.
<class 'pandas.core.frame.DataFrame'>
Int64Index: 580 entries, 4792 to 6385
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   year      580 non-null    float64
 1   activity  501 non-null    object 
 2   sex       529 non-null    object 
 3   age       77 non-null     object 
 4   cat_age   580 non-null    object 
 5   century   580 non-null    object 
dtypes: float64(1), object(5)
memory usage: 31.7+ KB


In [80]:
# Check the dataset for the 'XX' century
n_rows_cen20 = df_cen20.shape[0]
print(f'There are {n_rows_cen20} cases that happened during the XX century.')
df_cen20.info()

There are 3538 cases that happened during the XX century.
<class 'pandas.core.frame.DataFrame'>
Int64Index: 3538 entries, 2225 to 5803
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   year      3538 non-null   float64
 1   activity  3194 non-null   object 
 2   sex       3135 non-null   object 
 3   age       1867 non-null   object 
 4   cat_age   3538 non-null   object 
 5   century   3538 non-null   object 
dtypes: float64(1), object(5)
memory usage: 193.5+ KB


In [81]:
# Check the dataset for the 'XXI' century
n_rows_cen21 = df_cen21.shape[0]
print(f'There are {n_rows_cen21} cases that happened during the XXI century.')
df_cen21.info()

There are 2207 cases that happened during the XXI century.
<class 'pandas.core.frame.DataFrame'>
Int64Index: 2207 entries, 0 to 2224
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   year      2207 non-null   float64
 1   activity  2109 non-null   object 
 2   sex       2109 non-null   object 
 3   age       1650 non-null   object 
 4   cat_age   2207 non-null   object 
 5   century   2207 non-null   object 
dtypes: float64(1), object(5)
memory usage: 120.7+ KB


We are not even 1/4 of the 21st century and the number of shark incidents getting closer to the number of incidents in the 20th century. However, we also have to consider that there may be a probability that not all incidents were registered in the last century. 

## Gender

In [82]:
# Number of cases in the XIX century that happened involving each gender
df_cen19.sex.value_counts(dropna=False)

M      507
NaN     51
F       22
Name: sex, dtype: int64

In [83]:
# % of cases in the XIX century that happened involving each gender
(df_cen19.sex.value_counts(dropna=False) / n_rows_cen19) * 100

M      87.413793
NaN     8.793103
F       3.793103
Name: sex, dtype: float64

In [84]:
# Number of cases in the XX century that happened involving each gender
df_cen20.sex.value_counts(dropna=False)

M      2865
NaN     403
F       270
Name: sex, dtype: int64

In [85]:
# % of cases in the XX century that happened involving each gender
(df_cen20.sex.value_counts(dropna=False) / n_rows_cen20) * 100

M      80.977954
NaN    11.390616
F       7.631430
Name: sex, dtype: float64

In [86]:
# Number of cases in the XXI century that happened involving each gender
df_cen21.sex.value_counts(dropna=False)

M      1731
F       378
NaN      98
Name: sex, dtype: int64

In [87]:
# % of cases in the XXI century that happened involving each gender
(df_cen21.sex.value_counts(dropna=False) / n_rows_cen21) * 100

M      78.432261
F      17.127322
NaN     4.440417
Name: sex, dtype: float64

This result reflects each gender's behaviors in each century. Long time ago, women usually would just take care of the house and were not allowed to do many things.

## Activity

In [88]:
# Top 3 activities involved in shark incidentes during the XIX century - Number
df_cen19.activity.value_counts().head(3)

Swimming    96
Bathing     86
Fishing     42
Name: activity, dtype: int64

In [89]:
# Top 3 activities involved in shark incidentes during the XIX century - %
(df_cen19.activity.value_counts().head(3) / n_rows_cen19) * 100

Swimming    16.551724
Bathing     14.827586
Fishing      7.241379
Name: activity, dtype: float64

In [90]:
# Top 3 activities involved in shark incidentes during the XX century - Number
df_cen20.activity.value_counts().head(3)

Swimming    515
Surfing     360
Fishing     269
Name: activity, dtype: int64

In [91]:
# Top 3 activities involved in shark incidentes during the XX century - %
(df_cen20.activity.value_counts().head(3) / n_rows_cen20) * 100

Swimming    14.556246
Surfing     10.175240
Fishing      7.603166
Name: activity, dtype: float64

In [92]:
# Top 3 activities involved in shark incidentes during the XXI century - Number
df_cen21.activity.value_counts().head(3)

Surfing     677
Swimming    296
Fishing     147
Name: activity, dtype: int64

In [93]:
# Top 3 activities involved in shark incidentes during the XXI century - %
(df_cen21.activity.value_counts().head(3) / n_rows_cen21) * 100

Surfing     30.675125
Swimming    13.411871
Fishing      6.660625
Name: activity, dtype: float64

The result shows that for each century there was different activities and habits. For example, it was more common for people to bathe in the nature during the 19th century and the surfing just appeared in the 20th century.